### ISSUE TITLE
Release 4.2 - June 2021

### ISSUE THREAD

**Issue by philwo** [bazelbuild/bazel#13558]
# Status of Bazel 4.2

This release will use Bazel 4.1.0 as its baseline and we will apply selected cherry-picks and backports on top of it. Please request cherry-picks that you'd like to get into Bazel 4.2.0 here via a comment.

- Expected release date: June 2021
- [List of release blockers](https://github.com/bazelbuild/bazel/labels/Release%20blocker)

To report a release-blocking bug, please file a bug using the `Release blocker` label, and cc me.

Task list:

- [x] Pick release baseline: https://github.com/bazelbuild/bazel/commit/c4975efdd7ede7b46637bf353209d9ac371181a5
- [x] Create release candidate: https://releases.bazel.build/4.2.0/rc1/
- [x] Check downstream projects: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2105
- [x] [Create draft release announcement](https://docs.google.com/document/d/1wDvulLlj4NAlPZamdlEVFORks3YXJonCjyuQMUQEmB0/edit)
- [x] Send for review the release announcement PR: https://github.com/bazelbuild/bazel-blog/pull/258
- [x] Push the release, notify package maintainers: ...
- [x] Update the documentation
- [x] Push the blog post
- [x] Update the [release page](https://github.com/bazelbuild/bazel/releases/tag/4.2.0)

**Comment by philwo** [comment]
### Requested cherry picks

**Unreviewed:**
- [ ] https://github.com/bazelbuild/bazel/commit/8ba48ad869006d87abc84e2f699fbc59eb8470f8 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-827091633
- [ ] https://github.com/bazelbuild/bazel/commit/48eee8b4b447a2ad11df28dd81a2ccb65562b5f5 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-827164697
- [ ] https://github.com/bazelbuild/bazel/commit/763dd0ce6e1644bf895231432f616427a11d385a in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-830266538
- [ ] https://github.com/bazelbuild/bazel/commit/c22855857d3fb6c13ef70c48710793bbc4f195ea in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-830509694
- [ ] https://github.com/bazelbuild/bazel/commit/2d990cf2fc9607599486242c6c3d37f3d3b266ce in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-838501438
- [ ] https://github.com/bazelbuild/bazel/commit/5b95d9162b56b51c8e8f66258981ddf3c5d96765 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-839842570
- [ ] https://github.com/bazelbuild/bazel/commit/779d66019210f54e10a1343ee004df72a8dec812 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-839842570
- [ ] https://github.com/bazelbuild/bazel/commit/f4b5e0233341977aaa76593ca032d9ac4eba7444 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-839842570
- [ ] https://github.com/bazelbuild/bazel/commit/b2231c56d78c6d37bcb6f11e1e50fe68ee336b4a in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-839842570
- [ ] https://github.com/bazelbuild/bazel/commit/1a35dc5b51f4e1f1076f80b8fbe8bde517c5aa7a in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-839842570
- [ ] https://github.com/bazelbuild/bazel/commit/aaae8ce2881c8c1a5d4ad64f20d6e71aa372cf2e in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-840033103
- [ ] https://github.com/bazelbuild/bazel/commit/6dc941e58dfc1d4a9714a76b921fbe11fce658ed in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-843320116
- [ ] https://github.com/bazelbuild/bazel/commit/c9f9eedc313fc043755dd30ae7e01478eed7c417 in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-843380161
- [ ] https://github.com/bazelbuild/bazel/commit/2f0927a4fd9342f4dcfd43475d3f1c90c523584f in https://github.com/bazelbuild/bazel/issues/13505#issuecomment-846448423
- [ ] https://github.com/bazelbuild/bazel/commit/fd9cffdcaf05551126e66f1cd62815eaa1af6bd9 in https://github.com/bazelbuild/bazel/issues/13505#issuecomment-852355286
- [ ] https://github.com/bazelbuild/bazel/commit/42e05dbd606a5059126bda14cdfffa5e0653b652 in https://github.com/bazelbuild/bazel/issues/13505#issuecomment-854266569

**Accepted:**
- [ ] ...

**Rejected:**
- [ ] ...

**Comment by quval** [comment]
Would it be possible to cherry-pick 1e258d2a7a5221613047e5cee0aaec5b56045d2b, d0676693310215407224c1b8e8aea9e3eddc183d, f1e0d346c8235c855e61afc2adb870e4b895e002 and 8186fbb47ab964a9affa9a0fc6315fcdbde2b5aa again into 4.2.0, as well as the commits fixing #13459?

**Comment by thii** [comment]
Can I request these two? Some rulesets like rules_swift are pretty aggressive on using these new features.
- 50ce3f973cbc96a0326560a31b736a4f0ca8dc62: starlark: allow lambda expressions
- 5ca20643e54e5cff1eb2939d044f86f96861176a: starlark: allow nested def statements

**Comment by ahumesky** [comment]
Could we please have 80feea06cbf06486280dac065e3caec5facbedb9 for the android rules, thanks!

**Comment by ahumesky** [comment]
And also a followup 0e652737988e3c115e98e1552f6fada52bc2b9a2 for the android rules

**Comment by cpsauer** [comment]
Any chance we could cherrypick https://github.com/bazelbuild/bazel/commit/1a0285c3b64b121268ced3eb9ad9d5ba396b4905?

**Comment by gregestren** [comment]
Request:

 - https://github.com/bazelbuild/bazel/commit/64157a59d187f32201faa8503194b1cfa23b6f16 Respect Starlark options with values in removeStarlarkOptions() (to resolve https://github.com/bazelbuild/bazel/issues/12808#issuecomment-849128627).

**Comment by brentleyjones** [comment]
Can we also include https://github.com/bazelbuild/bazel/commit/7643d3abb654dec9cf7b49e2fce684dd45e36bc0?

**Comment by brentleyjones** [comment]
Also this one: https://github.com/bazelbuild/bazel/commit/95c793135c806601aa696c4541609fbee722f90c

**Comment by mai93** [comment]
Can you please include 1bca1bde7db06340559878a7d54b1f5698e7ddd6?

**Comment by mai93** [comment]
Also this one: f1b37b5d610728634d92ef4ebd4d1435c0c8605d

**Comment by brentleyjones** [comment]
Can we please include https://github.com/bazelbuild/bazel/commit/f0983dfddb85fd3f5b3c79afea8153dae4264746?

**Comment by martis42** [comment]
Please include https://github.com/bazelbuild/bazel/commit/cac82cf00649217e621e86825e9e99b3e56b0d34

**Comment by csmulhern** [comment]
Can we include b06f495?

**Comment by FaBrand** [comment]
Can this bugfix be included 382fd6453c0ff65940de4e1557ad8f17ad7198d7 ?

**Comment by meteorcloudy** [comment]
We'll need https://github.com/bazelbuild/bazel/commit/b4b0c321910bc968736ef48e8140528ea7d323cd to fix https://github.com/bazelbuild/bazel/issues/13514

**Comment by jablin** [comment]
Please include 5d7df56b88a9f5c51e1d84866d5548fc10cd2692 .

**Comment by philsc** [comment]
Can we get 9868455f017ddf00eaef02cb52551b45212d4136 and d052ececddf54587c576f876ce13d4b8f4aacb0b cherry-picked?
The former is required for the latter to apply cleanly.

**Comment by jesec** [comment]
Please consider: 

- bcce6dd026e90336e80616a8c1004a79a2f8640c (#12460)
- 1c29dff364fd34d7e6158c812cfd6d96f66be747 (#12683)
- 52b517f223e98e9c538fa945e43b7eecdc7c9cdc

Thanks.

**Comment by coeuvre** [comment]
Please include following commits for remote execution (in committed order):

- [ ] ae53991f2e Remote: Add RemoteExecutionService as a layer between spawn execution and remote execution.
- [ ] cd477f266e Pass only the top-level artifacts to `ExecutorLifecycleListener` and make the computation more efficient.
- [ ] 1e803abc29 Remove default build options.
- [ ] 275c54c982 Change the signature of tree artifact expansion to allow passing a reference to children from `TreeArtifactValue` to `ArtifactExpanderImpl`.
- [ ] 300c11ef5e Separate Fileset-symlink errors out from I/O exceptions, and propagate them everywhere.
- [ ] fd9cffdcaf Suppress interrupted status during pool closure
- [ ] 7643d3abb6 Remote: Register "remote" strategy even if remote execution is not available.
- [ ] 4ca8946a8e Remote: Add --experimental_capture_corrupted_outputs flag.
- [ ] f0983dfddb Remote: Report checking cache status before the action is scheduled to run remotely.
- [ ] 4261b2a119 Inclusive naming fixit
- [ ] ba5b2a7c94 when writing to local disk cache, open files later in order to avoid "too many open files"
- [ ] 97fb2cf193 Remote: Fix a bug that the XML generation is executed even if test.xml is generated when build with --remote_download_minimal.
- [ ] 2ace97f569 Move `ActionInputHelper` to a separate target.
- [ ] af42653e6f Automatic code cleanup.
- [ ] #13557
- [ ] #13594

**Comment by katre** [comment]
I plan to start assembling the branch for this release on Monday, July 12, so expect further updates around then.

It's a long list of cherrypicks so it may take some time to determine which can be applied and which cannot.

**Comment by jayconrod** [comment]
Could we please include c9f9eed? It will unblock some work on Go static analysis.

**Comment by cpsauer** [comment]
Any chance we could also consider the AAR proguard bugfix https://github.com/bazelbuild/bazel/commit/c8c0d94a49e1b865d95c6d245c2d152c7c7c9722

**Comment by cpsauer** [comment]
Could we consider https://github.com/bazelbuild/bazel/commit/c8c0d94a49e1b865d95c6d245c2d152c7c7c9722 to fix pro guards with aars?

**Comment by katre** [comment]
We are currently looking at 64 cherrypicks for this release, which is a huge number.

@coeuvre: Since #13594 does not appear to be merged, it cannot be cherrypicked. Is this releated to #13321, which is currently marked as a release blocker?

**Comment by katre** [comment]
In the first pass (with no attempt to clean up merges), 26 commits applied cleanly:
-  2d990cf2fc Fix #10127: Remove Python 2 dependency from tools/android.
-  64157a59d1 Respect Starlark options with values in `removeStarlarkOptions()`
-  c8c0d94a49 Export proguard specs from aar_import
-  1e258d2a7a Allow exec groups to inherit from the rule or other exec groups.
-  d067669331 Support execution constraints per exec group
-  f1e0d346c8 Clean up RuleContext to use a Table instead of a Map of Maps.
-  8186fbb47a Documentation for #13110
-  42e05dbd60 Update threshold for long path shortening to be MAX_PATH - 4
-  d052ececdd Expand target_compatible_with to all non-workspace rules
-  1a0285c3b6 Fix stripping of macOS loadable bundles
-  8ba48ad869 Java coverage: fix handling of external files
-  48eee8b4b4 [Bazel] Fix mobile-install for python2
-  763dd0ce6e Add `stub_shebang` to `py_runtime`
-  b2231c56d7 Move use of legacy sandbox -> local fallback to only be used after all strategies have been tried, and improve messages around it.
-  9b18d951a5 Rename ToolchainCollection.getExecGroups to getExecGroupNames.
-  6dc941e58d Remove restriction on generate_pdb_file to be only used in dbg and fastbuild mode
-  5b95d9162b Check the result of Future.cancel() when cancelling the other branch of dynamic execution.
-  aaae8ce288 Update DEFAULT_MACOS_CPU to match host
-  bcce6dd026 Add the TEMP_FAILURE_RETRY macro to linux-sandbox-pid1.cc.
-  2f0927a4fd Fix symlink creation on older Windows versions
-  b4b0c32191 Fix unix toolchain for macos arm64 platform
-  fd9cffdcaf Suppress interrupted status during pool closure
-  b06f49507e Fixes incorrect install names on darwin platforms
-  0c07c2e657 Remote: Add remoteCacheable key to execution log
-  af42653e6f Automatic code cleanup.
-  07a84ce31d Remote: Another attempt to fix the CancellationException error in AsyncTaskCache caused by a race condition.

The remaining 31 did not apply cleanly and will require further effort (possibly some of these depend on earlier ones and so will be fine when their required ancestor is applied):
-  5ca20643e5 starlark: allow nested def statements
-  50ce3f973c starlark: allow lambda expressions
-  c22855857d fix main repo starlark options parsing - now flags passed on the command line as --@main_workspace//flag and --//flag will both parse to --//flag. Before this CL, the former maintained its workspace prefix and we would get different entries for these two formats.
-  1c29dff364 Support mips64 and riscv64
-  1a35dc5b51 Force source files to be readable before copying them from sandbox.
-  f4b5e02333 Let workers finish lost races without delaying dynamic execution.
-  c9f9eedc31 Fix label_flag and label_setting to not have a dependency on the default
-  ae53991f2e Remote: Add RemoteExecutionService as a layer between spawn execution and remote execution.
-  cd477f266e Pass only the top-level artifacts to `ExecutorLifecycleListener` and make the computation more efficient.
-  dcceaa30c1 Split ExecGroup into a new target.
-  64534e2ded Move DEFAULT_EXEC_GROUP_NAME from ToolchainCollection to ExecGroup.
-  762b5d8407 BuildViewForTesting should directly call into ConfiguredTargetFunction.
-  1e803abc29 Remove default build options.
-  a116649162 Move exec group tests out of platforms_test and into integration.
-  afba8ac088 Update creating exec groups that explicitly copy from defaults.
-  779d660192 Only allow worker async finishing when sandboxed.
-  645c42bae4 Create a new ExecGroupCollection container to manage exec group inheritance and exec property parsing.
-  1bca1bde7d Add `required_providers` attribute to Starlark defined aspects.
-  275c54c982 Change the signature of tree artifact expansion to allow passing a reference to children from `TreeArtifactValue` to `ArtifactExpanderImpl`.
-  300c11ef5e Separate Fileset-symlink errors out from I/O exceptions, and propagate them everywhere.
-  52b517f223 third_party: replace "reactive_streams" and "rxjava3" in Debian build
-  80feea06cb Automated rollback of commit 9a1d428e33bfae1ec5b68250d4732b72346b8b39.
-  95c793135c Remote: Add remoteCacheHits to BEP BuildMetrics.
-  7643d3abb6 Remote: Register "remote" strategy even if remote execution is not available.
-  4ca8946a8e Remote: Add --experimental_capture_corrupted_outputs flag.
-  f0983dfddb Remote: Report checking cache status before the action is scheduled to run remotely.
-  4261b2a119 Inclusive naming fixit
-  ba5b2a7c94 when writing to local disk cache, open files later in order to avoid "too many open files"
-  97fb2cf193 Remote: Fix a bug that the XML generation is executed even if test.xml is generated when build with --remote_download_minimal.
-  2ace97f569 Move `ActionInputHelper` to a separate target.
-  0f812eb5e5 Remote: Display download progress when actions are downloading outputs from remote cache.

I have not yet pushed the release branch or started testing, I'm going to examine the remaining cherrypicks first.

**Comment by martis42** [comment]
@katre 
https://github.com/bazelbuild/bazel/commit/cac82cf00649217e621e86825e9e99b3e56b0d34 is missing on your list (or is this change disqualified due to some reason?)

**Comment by katre** [comment]
@martis42 I missed copying it into the list. I'll add it to the next set.

**Comment by FaBrand** [comment]
@katre could 382fd6453c0ff65940de4e1557ad8f17ad7198d7 (#13496) also be added to that set?

**Comment by philsc** [comment]
Any chance we could include 08936aecb96f2937c61bdedfebcf1c5a41a0786d? "Add external_include_paths C++ feature".

**Comment by katre** [comment]
@FaBrand That commit has not been merged to master, so I will not cherrypick it to the 4.2 release branch. Please get it merged to master first and I will evaluate whether it makes sense.

**Comment by coeuvre** [comment]
> @coeuvre: Since #13594 does not appear to be merged, it cannot be cherrypicked. Is this releated to #13321, which is currently marked as a release blocker?

It's a fix for #13349 and should be merged in a day or two.

IIRC, #13321 is a reminder that if we included 7149f57 in the release, we also need to include 24c980b, e6f1066, 4efeac9 and b863dbf2d2a327364483c72f1c5e73962af61148.

The reason that some commits for remote execution did not applied cleanly are probably they are depending on b863dbf2d2a327364483c72f1c5e73962af61148. We may need to include this series of commits.

**Comment by FaBrand** [comment]
katre I think i picked the wrong sha. a9e7577d3f96192114331408fb5f629fab5ba3db is the cherry-pick to master of #13496

**Comment by katre** [comment]
Okay, after further testing even some of the commits I thought had applied cleanly do not actually build, much less pass tests.

I'm working on a new process to streamline this, hopefully I can report back soon.

**Comment by illicitonion** [comment]
Please could I request a cherry-pick of https://github.com/bazelbuild/bazel/commit/20aff82091125f1d8dced5aa36069ef5ae730878 to allow for more precise determining which targets should be run in CI? Should apply very cleanly! :) Thanks!

**Comment by katre** [comment]
Okay, the current system is untenable, given the large number of requested cherrypicks and the wide divergence between the base of the LTS branch and the current Bazel master branch.

Therefore, here's the new process: I have created a release patch branch: https://github.com/bazelbuild/bazel/tree/release-4.2.0-patches

If you want a commit or commit series cherrypicked into the 4.2.0 release, create a PR against that branch and assign it to me for review. The PR description needs to have the following:
- A clear justification for the proposed cherrypick
- A list of the original commits that are being picked
- Ideally something that I can copy into the release notes about the cherrypick

I have two in flight now:
- https://github.com/bazelbuild/bazel/pull/13670: Most of the carryover requests from https://github.com/bazelbuild/bazel/issues/13558#issuecomment-855718849
- https://github.com/bazelbuild/bazel/pull/13671: A fix for Starlark options in bazelrc

If you have a cherrypick request, you have a few options:
1. Cherrypick the commits and create the PR yourself
2. Find a champion on the Bazel team (possibly the original commit author) to do this for you
3. Convince me that the request is sufficiently high value to try again to cherrypick the commit, effectively making me the champion.

We may or may not keep this system going forward: we'll see how it works out this cycle.

Once sufficient time has passed that I think most cherrypicks have been requested as PRs, I'll create an actual release candidate from the branch and we can resume the usual release testing.

I'll post a separate comment in a minute with the status of the current requests in this thread.

**Comment by katre** [comment]
Status of requested cherrypicks:

Carryover:
- 779d66019210f54e10a1343ee004df72a8dec812 - Doesn't apply cleanly. Someone who understands this (@larsrc-google ?) needs to examine the situation and see if this is a) worth it, and b) can be edited to be picked.

Exec group handling and "test" exec group, requested by @quval. I'll champion this, since I wrote the original code
- 1e258d2
- d067669
- f1e0d34
- 8186fbb
- #13459

Update to support rules_swift, @thii. This doesn't apply cleanly, fix it and send a PR.
- 50ce3f9
- 5ca2064

Android rules support, @ahumesky. This doesn't apply cleanly, fix it and send a PR.
- 80feea0
- 0e65273

@cpsauer requested this. I don't know why, and they don't apply cleanly.
- 1a0285c

@brentleyjones requested this. I don't know why, and they don't apply cleanly.
- 7643d3a
- f0983df

@mai93 requested this. I don't know why, and they don't apply cleanly.
- 1bca1bd

@martis42 requested this. I don't know why, and they don't apply cleanly.
- cac82cf

@csmulhern requested this. I don't know why, and they don't apply cleanly.
- b06f495

@FaBrand requested this. I don't know why, and they don't apply cleanly.
- a9e7577

Fix #13514, @meteorcloudy. Doesn't apply cleanly.
- b4b0c32

@jablin requested this. I don't know why, and they don't apply cleanly.
- 5d7df56

Support target_compatible_with, @philsc. Doesn't apply cleanly.
- d052ece

Alpine, riscv64, other?, @jesec. This needs some updates to work with the older distfiles, as I recall.
- bcce6dd
- 1c29dff
- 52b517f

Remote exec fixes, @coeuvre. This can all be one PR, but needs a lot of work to merge.
- b863dbf - maybe?
- ae53991
- cd477f2
- 1e803ab
- 275c54c
- 300c11e
- fd9cffd
- 0c07c2e
- 95c7931
- 7643d3a
- 4ca8946
- f0983df
- 4261b2a
- ba5b2a7
- 97fb2cf
- 2ace97f
- af42653
- 07a84ce
- 0f812eb
-  #13594 Remote: Do not upload empty output to remote cache. #13594

Label dep cycle fixes, @jayconrod. I'll champion this as well.
- c9f9eed

Fix proguards with aars, @cpsauer. Doesn't apply cleanly.
- c8c0d94

external_include_paths, @philsc. Doesn't apply cleanly.
- 08936ae

**Comment by quval** [comment]
@katre Thanks for championing the exec group commits (and the whole release effort)! I was actually about to start cherry-picking these into a branch between your two posts – so if this would help, please let me know and I'll send a PR.

**Comment by katre** [comment]
@quval If you have it started, by all means send a PR. I'm concerned that either not every commit needed is in the issue, or that it just builds on other work I did on the StarlarkToolchainContext, because I ran in to all sorts of trouble yesterday trying to figure out what commits were needed.

**Comment by katre** [comment]
One further request: please indicate in the PR title that it's a cherrypick request, and please link to this issue for tracking purposes. Thanks!

**Comment by quval** [comment]
@katre I tried, but it's proving hard. Cherry-picking 645c42bae42fc8ea1c5b7c53e4962f1987ff888e seems to require unravelling – or cherry-picking – a lot of work. You probably have much more context than I to get this right. I will, however, say that we can live without this change in Bazel 4.2.0. We might try the rolling releases instead (or wait for 5.0).

**Comment by martis42** [comment]
@katre In my opinion it is a bug that `config_setting` allows setting the `visibility` attribute, but it has no effect. https://github.com/bazelbuild/bazel/commit/cac82cf00649217e621e86825e9e99b3e56b0d34 resolves this bug. This helps reducing confusion on developer side and allows limiting where `config_setting` is used in the stack to prevent it being spreading into domains for which it was never intended.

How exactly is the process for cherry-picking to 4.2 defined. Is a clean application of the cherry-pick a must, or can we still provide a PR with the cherry-pick, even if a merge conflict has to be resolved?

**Comment by katre** [comment]
@martis42: Please send me a PR with the justification, bug link, and original commit, as I described earlier. The commit in the PR should start from "git cherry-pick <original commit>", but if you need to edit files to get them to merge that is fine. I just want to avoid including entirely new commits in an LTS release.

**Comment by gregestren** [comment]
@martis42 unfortunately there's also https://github.com/bazelbuild/bazel/commit/36d228bd792f4332c7486c4e5f9c78e4b55f4b06 - existing builds depend on the incorrect functionality, which makes https://github.com/bazelbuild/bazel/commit/cac82cf00649217e621e86825e9e99b3e56b0d34 an incompatible change.
We probably want https://github.com/bazelbuild/bazel/commit/79989f9becc2edefe8b35f7db687bf8de03e3580 instead, which rolls forward behind a feature flag.

**Comment by martis42** [comment]
@gregestren Thanks for the clarification. Sadly https://github.com/bazelbuild/bazel/commit/79989f9becc2edefe8b35f7db687bf8de03e3580 has a serious merge conflict which I am not able to resolve. Seems like we have to wait for Bazel 5.0.

**Comment by katre** [comment]
Thanks everyone for the help with merging the cherrypicks. Based on the current rate of incoming PRs I'll cut RC1 on Monday morning. If you get one in after that, it's fine, I can add it to RC2 afterwards.

**Comment by katre** [comment]
Full list of cherrypick PRs so far (since not all are linked to this issue), for use in release notes:
- #13670
- #13671
- #13675
- #13677
- #13683
- #13686
- #13688

Plus @larsrc-google's in-progress cherrypicks for worker updates.

**Comment by ZacxDev** [comment]
@katre I would love to see https://github.com/bazelbuild/bazel/commit/e5c53cf23815641608aed530a8185092ed3e2690 included in 4.2.0 so we can compile on alpine.  I wasn't the original author but I'm happy to send a cherry-pick PR if it is ready to go in.

**Comment by katre** [comment]
@ZacxDev: I have no idea where that commit comes from: it's not in the current Bazel master branch. Can you identify the commit from Bazel/master that should be cherrypicked? Once you do, feel free to send a PR for the merge.

**Comment by ZacxDev** [comment]
@katre My bad, I think this is the one that is on master: https://github.com/bazelbuild/bazel/commit/bcce6dd026e90336e80616a8c1004a79a2f8640c

**Comment by gregestren** [comment]
@martis42 - trying to support you! Please see https://github.com/bazelbuild/bazel/pull/13701.

**Comment by ahumesky** [comment]
Thanks @katre, I've sent
https://github.com/bazelbuild/bazel/pull/13702
https://github.com/bazelbuild/bazel/pull/13703

**Comment by cpsauer** [comment]
Have filed my two as PRs. (Backlinked just above.) 
Seemed like both just applied cleanly after all?

**Comment by katre** [comment]
@cpsauer It's possible they depended on other things that were merged, and so now they applied cleanly.

I'll get these (and #13701) merged today and cut RC1.

**Comment by katre** [comment]
Created and pushed the RC:

```
scripts/release/release.sh create --force_rc=1 4.2.0 release-4.2.0-patches
scripts/release/release.sh push
```

Waiting for tests to finish.

**Comment by katre** [comment]
Bazel 4.2.0rc1 is now available for those that want to try it out.

You can download it from: https://releases.bazel.build/4.2.0/rc1/index.html

Please report regressions to https://github.com/bazelbuild/bazel/issues as soon as possible.

**Comment by brentleyjones** [comment]
https://github.com/bazelbuild/bazel/issues/13711

**Comment by katre** [comment]
Looks like potentially three release blockers:
- #13711
- #13712
- #13713

I'll try to bisect and identify failures for the last two, there's current no repro instructions for #13711.

**Comment by katre** [comment]
Full downstream test run: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2105

**Comment by katre** [comment]
List of cherrypicks for an eventual RC2:
- #13714

**Comment by Bencodes** [comment]
This would be a good fix to get into 4.2 https://github.com/bazelbuild/bazel/pull/13653

**Comment by konste** [comment]
While testing 4.2.0rc1 I noticed that the maximum number of actions run concurrently is now 40, while with Bazel 4.1.0 it was 80. We use machines with 40 cores and HyperThreading which is usually seen as 80 virtual cores. I doubt the issue is anything more than cosmetic change, but still curious which cherry pick could have caused that and what actually changed. And I double-checked that the issue is not imaginary and only connected to the version of Bazel.

**Comment by katre** [comment]
@Bencodes Once that is submitted you can feel free to submit a PR for the cherrypick against the release-4.2.0-patches branch.

**Comment by katre** [comment]
@konste Please submit an issue, ideally with a reproduction, so that we can evaluate it and decide whether it is a release blocker.

**Comment by konste** [comment]
https://github.com/bazelbuild/bazel/issues/13726 does not seem to be a regression from 4.1.0 to 4.2.0, but still looks problematic enough to be a release blocker for 4.2.0.

**Comment by katre** [comment]
As all release blockers have been addressed, I'll cut RC2 later today.

**Comment by katre** [comment]
RC2 is created and available at https://releases.bazel.build/4.2.0/rc2/index.html

**Comment by katre** [comment]
Failures from https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118:

- Android Studio Plugin
  - Fails at HEAD
  - Missing input file
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#6f213797-e732-4e8a-8f00-8e494f762664

- Android Testing
  - Fails at HEAD
  - Requires Android build tools 30.0.0, 28.0.2 provided
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#9559ba81-a9cd-4e0c-bb8d-26cba7202083

- Bazel Examples
  - Expects version to be an int
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#13665dfe-5551-4509-89c5-5e13c17f11e8

- Bazelisk
  - Fails at HEAD
  - Test failure: https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/51c9e713-c33f-4264-a68b-93629b828c29/go_bazelisk_test/test.log
  - `FAIL: 'bazelisk version' of an unreleased binary must not print a build label.`

- Intellij Plugin Aspect
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#5920688d-f90e-437b-8ee5-44f4330d2752

- rules_gwt
  - Fails at HEAD
  - Issue with Ubuntu 18.04
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#efac61ba-76ac-419b-ba44-ee37fe9ddc06

- rules_jvm_external
  - Fails at HEAD
  - Android tools version 30.0.0
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#2c117ff1-c20a-4472-a63f-cf5d9274e68d

- rules_nodejs
  - Fails at HEAD
  - https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2118#5ac3884a-1de9-41c2-99ce-b4705ddd275b

Nothing appears major, I'm going to keep investigating.

**Comment by katre** [comment]
List of cherrypicks for the eventual RC3:
- #13768

**Comment by katre** [comment]
@larsrc-google says there are a few more fixes for remote/dynamic workers, once those are in I think it'll be time to close this for cherrypicks and finalize the release.

Thanks everyone for working with the process!

**Comment by katre** [comment]
Barring new failures, I think it's time to close this release to cherrypicks. I'll create a new RC shortly.

**Comment by katre** [comment]
RC3 is created and available at https://releases.bazel.build/4.2.0/rc3/index.html

**Comment by katre** [comment]
Downstream tests have finished: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2125

- Android Testing (all)
  - Android build tools version 30.0
  - Same as HEAD

- Bazel (Ubuntu 20.02 (OpenJDK 11))
  - //src/test/java/com/google/devtools/build/lib/concurrent:ConcurrentTests failure
    - https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/143780c4-1780-4dfc-849c-56414b5743c5/src/test/java/com/google/devtools/build/lib/concurrent/ConcurrentTests/attempt_1.log
    - Filed #13789 to track triage

- Bazel Examples (several)
  - Expects bazel version to be an int
  - Known issue

- rules_gwt (Ubuntu 18.04)
  - Known failure - unsupported commands
  - Same as HEAD

- rules_jvm_external (several)
  - Android build tools version 30.0
  - Same as HEAD

- rules_nodejs (RBE Ubuntu 16.04)
  - Error: Cannot find module 'npm_directory_artifacts/node_modules/protractor/bin/protractor'.
  - Same as HEAD

Only one new release blocker!

**Comment by katre** [comment]
No new release blockers with RC3.

I'll plan to release this Thursday if no new blockers are found.

**Comment by katre** [comment]
No new blockers, but I'm still getting the release notes into order, so release today may not happen.

**Comment by katre** [comment]
Aiming for release early monday, release notes are now in final review.

**Comment by katre** [comment]
I am now going through internal release processes. Unfortunately this will add a few days until the release can be finalized. Please continue to test the latest RC during this period.

**Comment by alexeagle** [comment]
General question about bazel's release process: is the Google Container Registry ever going to get updates?
https://docs.bazel.build/versions/main/bazel-container.html references
`l.gcr.io/google/bazel:latest`

which is from last September

```
$ docker run --rm l.gcr.io/google/bazel:latest version
Extracting Bazel installation...
Build label: 3.5.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Sep 2 20:11:46 2020 (1599077506)
Build timestamp: 1599077506
Build timestamp as int: 1599077506
```

**Comment by ZacxDev** [comment]
> General question about bazel's release process: is the Google Container Registry ever going to get updates?
> https://docs.bazel.build/versions/main/bazel-container.html references
> `l.gcr.io/google/bazel:latest`
> 
> which is from last September
> 
> ```
> $ docker run --rm l.gcr.io/google/bazel:latest version
> Extracting Bazel installation...
> Build label: 3.5.0
> Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
> Build time: Wed Sep 2 20:11:46 2020 (1599077506)
> Build timestamp: 1599077506
> Build timestamp as int: 1599077506
> ```

Funny, I was looking for this exact thing last night.  They are planning on adding a container to the release artifacts but haven't gotten to it yet, tracked here: https://github.com/bazelbuild/continuous-integration/issues/1060

**Comment by katre** [comment]
@alexeagle (and @ZacxDev) Thanks for pointing this out. Once the release is finalized I'll reach out to the owners of `l.gcr.io/google/bazel:latest` and see if they can manually update, since bazelbuild/continuous-integration#1060 isn't done yet.

**Comment by katre** [comment]
PRs for RC4:
- #13846

**Comment by katre** [comment]
With those last two PRs in, I'm going to create RC4 this afternoon and aim to release next week (probably early Wednesday). Thanks everyone for your patience.

**Comment by katre** [comment]
RC4 is created and available at https://releases.bazel.build/4.2.0/rc4/index.html

**Comment by katre** [comment]
Downstream tests have finished: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2137

All failures are expected or match failures in the latest nightly.

**Comment by katre** [comment]
Release 4.2.0 is now live at https://github.com/bazelbuild/bazel/releases/tag/4.2.0.

Can @vbatts @petemounce and @excitoon update their packages? Thanks

**Comment by katre** [comment]
Blog post is live: https://blog.bazel.build/2021/08/18/bazel-4.2.html

**Comment by vbatts** [comment]
On 18/08/21 06:10 -0700, John Cater wrote:
>Release 4.2.0 is now live at https://github.com/bazelbuild/bazel/releases/tag/4.2.0.

Build still fails like 4.1.0 did
https://github.com/bazelbuild/bazel/pull/13534

I'll try their patch (and maybe carry that patch for a little while)

**Comment by vbatts** [comment]
after patching a couple spots, it's also failing on the same issue in external/com_google_absl/absl/synchronization/internal/graphcycles.cc which is not so straightforward to patch:

```shell
  /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.d '-frandom-seed=bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.o' -D__CLANG_SUPPORT_DYN_ANNOTATION__ -iquote external/com_google_absl -iquote bazel-out/k8-opt/bin/external/com_google_absl -Wall -Wextra -Wcast-qual -Wconversion-null -Wmissing-declarations -Woverlength-strings -Wpointer-arith -Wunused-local-typedefs -Wunused-result -Wvarargs -Wvla -Wwrite-strings -Wno-missing-field-initializers -Wno-sign-compare -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_absl/absl/synchronization/internal/graphcycles.cc -o bazel-out/k8-opt/bin/external/com_google_absl/absl/synchronization/_objs/graphcycles_internal/graphcycles.o)
Execution platform: //:default_host_platform
external/com_google_absl/absl/synchronization/internal/graphcycles.cc: In member function 'void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)':
external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:26: error: 'numeric_limits' is not a member of 'std'
  451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
      |                          ^~~~~~~~~~~~~~
external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:49: error: expected primary-expression before '>' token
  451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
      |                                                 ^
external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:52: error: '::max' has not been declared; did you mean 'std::max'?
  451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
      |                                                    ^~~
      |                                                    std::max
In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/algorithm:62,
                 from external/com_google_absl/absl/synchronization/internal/graphcycles.cc:38:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
 3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
```

this is on gcc-g++ version:
```shell
/usr/bin/g++ --version
g++ (GCC) 11.2.1 20210728 (Red Hat 11.2.1-1)
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by davido** [comment]
> after patching a couple spots, it's also failing on the same issue in external/com_google_absl/absl/synchronization/internal/graphcycles.cc which is not so straightforward to patch [...]

I bumped `absl` and `gRPC` versions in this [PR](https://github.com/bazelbuild/bazel/pull/13536) and it fixed it. Unfortunately, it broke the build on ancient OS versions, see this tracking [issue](https://github.com/bazelbuild/bazel/issues/13561).

I know, this is invasive change, but none come up with a less intrusive change yet.

**Comment by fmeum** [comment]
Bazel 4.2.0 introduced a regression in the handling of Starlark flags: In a transition, the value of such a flag cannot be read if the flag is referred to as `@repository_name//:flag_name` instead of `//:flag_name`, which as far as I know is required for that transition to be usable by other workspaces.

I noticed this while working on [rules_fuzzing](https://github.com/bazelbuild/rules_fuzzing), whose tests no longer pass with Bazel 4.2.0. Usage of `rules_fuzzing` by other workspaces does not seem to be affected.

I [created a reproducer](https://gist.github.com/fmeum/4e2b5d08aa9a7ed8df39e4362f4bfa3b) that prints 

- `regressed`with
  - Bazel 4.2.0
- `not_regressed` with
  - Bazel 4.0.0
  - Bazel 4.1.0
  - Bazel 5.0.0-pre.20210810.4
  - Bazel 4.2.0 with 80c59dea59 reverted

Unfortunately 98d376faeb206f14838156ce4cb305ddbfce08fa cannot be cleanly cherry-picked on top of Bazel 4.2.0 by itself, so reverting 80c59dea59 appears to be the safest way to mitigate this regression.

**Comment by katre** [comment]
@fmeum Please file this as a separate issue so that we can triage it.

**Comment by fmeum** [comment]
@katre Done: https://github.com/bazelbuild/bazel/issues/13890

**Comment by katre** [comment]
Okay, preparing to start a 4.2.1 patch release to deal with the regression.

**Comment by katre** [comment]
Cherrypicks for 4.2.1rc1:
- #13893

**Comment by katre** [comment]
[4.2.1rc1](https://releases.bazel.build/4.2.1/rc1/index.html) is now available for testing.

**Comment by alexeagle** [comment]
@katre thanks for the release! ping on

> I'll reach out to the owners of l.gcr.io/google/bazel:latest and see if they can manually update

**Comment by katre** [comment]
@alexeagle I spoke to the previous team, who have handed it off to @philwo, who is still trying to automate it.

In the meantime, the directions are apparently here: https://github.com/bazelbuild/bazel/issues/12367#issuecomment-729630805

Sorry for the confusion.

**Comment by vbatts** [comment]
> I bumped `absl` and `gRPC` versions in this [PR](https://github.com/bazelbuild/bazel/pull/13536) and it fixed it. Unfortunately, it broke the build on ancient OS versions, see this tracking [issue](https://github.com/bazelbuild/bazel/issues/13561).
> 
> I know, this is invasive change, but none came up with a less intrusive change yet.

thanks for this @davido. This will even require some rebasing to work with the dist zip of the release. 

I think I am hitting a personal wall to churn package updates for Fedora/Centos. I haven't used bazel in years, and was only spinning builds as a best-effort help. Would love to hand over https://copr.fedorainfracloud.org/coprs/vbatts/ and https://github.com/vbatts/copr-build-bazel to someone in the project.

**Comment by katre** [comment]
Downstream tests finished: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2152

All failures are either in the latest nightly or are expected (Bazel Examples).

**Comment by katre** [comment]
@vbatts Thanks for all your help keeping this working, we definitely appreciate it, but if this is getting too much work we understand.

Unfortunately I don't think core Bazel devs can take this over, but if someone else in the community finds the Fedora/Centos releases useful and wants to take them over, we're happy to assist.

**Comment by katre** [comment]
Bazel 4.2.1 is now live: https://github.com/bazelbuild/bazel/releases/tag/4.2.1

Please update your packages if possible: @vbatts @petemounce @excitoon

**Comment by anthpi** [comment]
When building 4.2.0 and 4.2.1 I get "tools/jdk/BUILD:41:15: //tools/jdk:dummy_toolchain: missing value for mandatory attribute 'java_runtime' in 'java_toolchain' rule" error. Any idea what could cause it?

**Comment by katre** [comment]
@anthpi Please file a separate issue with reproduction instructions and we'll investigate.

**Comment by omarzl** [comment]
Hello there! These are my first contributions to the repo, I sent a cherry pick pr #14086 to 4.2.1 but it is already released, can a 4.2.2 or 4.3.0 be opened? Or should I send it to 5.0? Thanks!

**Comment by limdor** [comment]
@katre there is a fix for #14178 that went in with https://github.com/bazelbuild/bazel/commit/063b5c9c2c09b4794010b9a169b44890ffc79ec4
I just tried to cherry pick this commit on top of the 4.2.1 tag and it is applied cleaned.
Could you and the Bazel team consider a 4.3 release and answer here about the decision?
Thanks

**Comment by meteorcloudy** [comment]
We are doing a 4.2.2 patch release for #14178 and probably https://github.com/bazelbuild/bazel/issues/14273.

**Comment by limdor** [comment]
> We are doing a 4.2.2 patch release for #14178 and probably #14273. Please send PRs for the fixes to https://github.com/bazelbuild/bazel/tree/release-4.2.2-patches

Greate, I created a PR for #14178, if I should add some special tag just let me know https://github.com/bazelbuild/bazel/pull/14318 currently tests are running

**Comment by bjacklyn** [comment]
Please consider #13985 as well -- fixes bazel UI bug introduced in 4.2.0

**Comment by meteorcloudy** [comment]
@bjacklyn Can you also send a PR to the `release-4.2.2-patches` branch?

**Comment by meteorcloudy** [comment]
Bazel 4.2.2rc1 is now available: https://releases.bazel.build/4.2.2/rc1/index.html

Cherry-picks since 4.2.1:
* 039c72e5e25a426cbf45c582bf325dcd112af4bd
* af74287f125b93119415ba35429b8638d7a986ea
* 748138f3d993d6ac3a1e2357865b4c7737629289
* dd8fe7719809f5731f3fc824464ff0c930cd5053

Downstream tests haven't show any breakage from the Bazel side: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2262

**Comment by meteorcloudy** [comment]
Bazel 4.2.2 is now live: https://github.com/bazelbuild/bazel/releases/tag/4.2.2

Please update your packages if possible: @vbatts @petemounce @excitoon

**Comment by petemounce** [comment]
Published!
